### ISSUE TITLE
magit-commit-add-log on file results in Wrong type argument: integer-or-marker-p, nil

### ISSUE THREAD

**Issue by jabranham** [magit/magit#4062]
Pressing `C` (`magit-commit-add-log`) on a file name in the magit-diff buffer (after the line `1 file changed, X insertions(+) and Y deletions(-)` results in wrong-type-argument. I'd expect it to insert `* filename:` into the commit edit message buffer. This used to work in previous magit so apologies if I missed something in the release notes.

1. Navigate to a git dir
2. `M-x magit-status
3. Stage a file with `s`, then `c c`
4. Switch to the `magit-diff: foo` buffer, navigate to the second line and press `C`

```
Debugger entered--Lisp error: (wrong-type-argument integer-or-marker-p nil)
  goto-char(nil)
  (save-restriction (widen) (goto-char pos) (funcall magit-commit-add-log-insert-function log (magit-file-relative-name) (and hunk (add-log-current-defun))))
  (save-excursion (save-restriction (widen) (goto-char pos) (funcall magit-commit-add-log-insert-function log (magit-file-relative-name) (and hunk (add-log-current-defun)))))
  (save-current-buffer (set-buffer buf) (save-excursion (save-restriction (widen) (goto-char pos) (funcall magit-commit-add-log-insert-function log (magit-file-relative-name) (and hunk (add-log-current-defun))))))
  (progn (if log nil (if (magit-commit-assert nil) nil (user-error "Abort")) (magit-commit-create) (while (not (setq log (magit-commit-message-buffer))) (sit-for 0.01))) (save-current-buffer (set-buffer buf) (save-excursion (save-restriction (widen) (goto-char pos) (funcall magit-commit-add-log-insert-function log (magit-file-relative-name) (and hunk (add-log-current-defun)))))))
  (let ((pos x606) (buf x604)) (progn (if log nil (if (magit-commit-assert nil) nil (user-error "Abort")) (magit-commit-create) (while (not (setq log (magit-commit-message-buffer))) (sit-for 0.01))) (save-current-buffer (set-buffer buf) (save-excursion (save-restriction (widen) (goto-char pos) (funcall magit-commit-add-log-insert-function log (magit-file-relative-name) (and hunk (add-log-current-defun))))))))
  (let* ((hunk (and (magit-section-match 'hunk) (magit-current-section))) (log (magit-commit-message-buffer)) (val (magit-diff-visit-file--noselect)) (x604 (car val)) (x605 (cdr val)) (x606 (car x605)) (x607 (cdr x605))) (let ((pos x606) (buf x604)) (progn (if log nil (if (magit-commit-assert nil) nil (user-error "Abort")) (magit-commit-create) (while (not (setq log (magit-commit-message-buffer))) (sit-for 0.01))) (save-current-buffer (set-buffer buf) (save-excursion (save-restriction (widen) (goto-char pos) (funcall magit-commit-add-log-insert-function log (magit-file-relative-name) (and hunk ...))))))))
  magit-commit-add-log()
  funcall-interactively(magit-commit-add-log)
  call-interactively(magit-commit-add-log nil nil)
  command-execute(magit-commit-add-log)

```

**Comment by kyleam** [comment]
> This used to work in previous magit [...]

Looks like the change happened with e9de5a43 (Use magit-diff-visit--noselect when appropriate, 2019-04-22).  That switched `magit-commit-add-log` to calling `magit-diff-visit-file--noselect` but didn't account for the possibility of nil being returned for the position.

There are currently only a few callers of `magit-diff-visit-file--noselect`: `magit-diff-visit-file--internal`, `magit-commit-add-log`, `magit-add-change-log-entry`, and `magit-diff-trace-definition`.  The first handles a position of nil fine, but the rest do not.  Given all the remaining commands go through `magit--with-temp-position`, I think the simplest fix would be

```diff
diff --git a/lisp/magit-utils.el b/lisp/magit-utils.el
index 8177a7e2..18084d0a 100644
--- a/lisp/magit-utils.el
+++ b/lisp/magit-utils.el
@@ -1187,7 +1187,7 @@ (defmacro magit--with-temp-position (buf pos &rest body)
      (save-excursion
        (save-restriction
          (widen)
-         (goto-char ,pos)
+         (goto-char (or ,pos 1))
          ,@body))))
 
 ;;; _
```

**Comment by kyleam** [comment]
> I ran into the same issue (with `magit-diff-trace-definition`) a few hours ago and instead of assuming a location that the user did not actually specify (as you are proposing), I decided to instead make the command unavailable in situations where it would currently error. See the `fix-4062` branch.

For `magit-diff-trace-definition` I agree that what you suggest is an improvement, though in nearly all realistic cases, where there is no function definition at the first line of the file, I think the call would abort with a reasonable "No function at point found".

For `magit-commit-add-log` and `magit-add-change-log-entry`, I'm not so sure it's an improvement.  I think it's probably more useful to add the file name alone than to error out (though I personally do not care either way).

> But its not just the buffer position that might be unknown, the file/buffer might also be undefined, so the solution should probably involve `(if (and buf pos) ... (user-error ...))`.

IIUC the primary way for a file/buffer to be unknown is that there's no file at point.  `magit-diff-visit-file--noselect` calls `magit-file-at-point` with a non-nil `ASSERT`, so it will signal a user-error in that case.

**Comment by tarsius** [comment]
> I think it's probably more useful to add the file name alone than to error out

Yes, provided we can still use `add-log.el` to do that.

**Comment by kyleam** [comment]
> Yes, provided we can still use `add-log.el` to do that.

With the fallback to 1 for a nil position, `magit-add-change-log-entry` (so `add-change-log-entry`) on a `[file diffstat diffbuf]` section seems to do that (that is, add just the file name to the changelog entry), producing something like

```
2020-03-06  [...]

	* lisp/magit-bisect.el: 
```

But I don't personally feel strongly about this or have any use for it, so I'd say we should go with your proposal.  If someone else ends up caring, they can make the case for that functionality.

---

When testing that, I noticed a problem when `magit-diff-visit-file--noselect` returns a blob-visiting buffer.  In that case, `add-change-log-entry` doesn't know how to get the file name.  That's not a problem specific to the `diffstat` sections; here's an example of what would be produced from a `[hunk file ...]` section:

```
2020-03-06  [....]

	*  (magit-bisect-run): 
```

I've pushed 1916e83a (magit-add-change-log-entry: Tell add-log about magit-buffer-file-name) to address that issue.
